# Módulo 07, Tarefa 01

Vamos começar a mexer na nossa base de projeto? Já fizemos um exercício de montar a variável resposta, já pudemos perceber que essa atividade pode não ser trivial. Vamos agora trabalhar a base para que fique propícia ao *scikitlearn* para trabalharmos.

Lembrando, a base se chama demo01.csv, e originalmente está publicada [aqui](https://www.kaggle.com/rikdifos/credit-card-approval-prediction).

#### 1) Carregue a base e avalie:

- As variáveis
- Tipos de dados de cada variável
- Quantidade de missings
- Distribuição da variável resposta (mau)

In [208]:
import pandas as pd
import matplotlib.pyplot as plt

pd.set_option('display.max_columns', None)
df = pd.read_csv(r'demo01.csv', sep = ',')

In [209]:
df.head(6)

sexo posse_de_veiculo posse_de_imovel  qtd_filhos            tipo_renda  \
0    M                Y               Y           0               Working   
1    F                N               Y           0  Commercial associate   
2    F                N               Y           0  Commercial associate   
3    M                Y               Y           0               Working   
4    F                Y               N           0               Working   
5    F                Y               N           0               Working   

                        educacao          estado_civil    tipo_residencia  \
0  Secondary / secondary special               Married  House / apartment   
1  Secondary / secondary special  Single / not married  House / apartment   
2  Secondary / secondary special  Single / not married  House / apartment   
3               Higher education               Married  House / apartment   
4              Incomplete higher               Married  House / apartment   
5              Incomplete higher               Married  House / apartment   

       idade  tempo_emprego  possui_celular  possui_fone_comercial  \
0  58.832877       3.106849               1                      0   
1  52.356164       8.358904               1                      0   
2  52.356164       8.358904               1                      0   
3  46.224658       2.106849               1                      1   
4  29.230137       3.021918               1                      0   
5  29.230137       3.021918               1                      0   

   possui_fone  possui_email  qt_pessoas_residencia    mau  
0            0             0                    2.0  False  
1            1             1                    1.0  False  
2            1             1                    1.0  False  
3            1             1                    2.0  False  
4            0             0                    2.0  False  
5            0             0                    2.0  False

* Informações sobre tamanho e tipo de dados do Data Frame

In [210]:
df.shape

(16650, 16)

In [211]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16650 entries, 0 to 16649
Data columns (total 16 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   sexo                   16650 non-null  object 
 1   posse_de_veiculo       16650 non-null  object 
 2   posse_de_imovel        16650 non-null  object 
 3   qtd_filhos             16650 non-null  int64  
 4   tipo_renda             16650 non-null  object 
 5   educacao               16650 non-null  object 
 6   estado_civil           16650 non-null  object 
 7   tipo_residencia        16650 non-null  object 
 8   idade                  16650 non-null  float64
 9   tempo_emprego          16650 non-null  float64
 10  possui_celular         16650 non-null  int64  
 11  possui_fone_comercial  16650 non-null  int64  
 12  possui_fone            16650 non-null  int64  
 13  possui_email           16650 non-null  int64  
 14  qt_pessoas_residencia  16650 non-null  float64
 15  ma

* Checando possíveis dados faltantes

In [212]:
df.isna().sum()

sexo                     0
posse_de_veiculo         0
posse_de_imovel          0
qtd_filhos               0
tipo_renda               0
educacao                 0
estado_civil             0
tipo_residencia          0
idade                    0
tempo_emprego            0
possui_celular           0
possui_fone_comercial    0
possui_fone              0
possui_email             0
qt_pessoas_residencia    0
mau                      0
dtype: int64

* Removendo dados duplicados e resetando o index

In [213]:
df.drop_duplicates(inplace = True)

In [214]:
df.reset_index(inplace = True)
df.tail()

index sexo posse_de_veiculo posse_de_imovel  qtd_filhos  \
6765  16644    F                N               Y           1   
6766  16645    F                N               Y           0   
6767  16646    F                N               Y           0   
6768  16647    M                Y               Y           0   
6769  16649    F                N               Y           0   

                tipo_renda                       educacao  \
6765               Working  Secondary / secondary special   
6766               Working  Secondary / secondary special   
6767  Commercial associate  Secondary / secondary special   
6768               Working  Secondary / secondary special   
6769             Pensioner               Higher education   

              estado_civil    tipo_residencia      idade  tempo_emprego  \
6765  Single / not married  House / apartment  34.857534       3.101370   
6766        Civil marriage  House / apartment  54.109589       9.884932   
6767               Married  House / apartment  43.389041       7.380822   
6768               Married  House / apartment  30.005479       9.800000   
6769               Married  House / apartment  33.936986       3.630137   

      possui_celular  possui_fone_comercial  possui_fone  possui_email  \
6765               1                      1            1             0   
6766               1                      0            0             0   
6767               1                      1            1             0   
6768               1                      1            0             0   
6769               1                      0            1             1   

      qt_pessoas_residencia   mau  
6765                    1.0  True  
6766                    2.0  True  
6767                    2.0  True  
6768                    2.0  True  
6769                    2.0  True

* Distribuição da variável 'mau'

In [215]:
df['mau'].value_counts()

False    6472
True      298
Name: mau, dtype: int64

In [216]:
round(df['mau'].value_counts(normalize = True), 2) * 100

False    96.0
True      4.0
Name: mau, dtype: float64

In [217]:
pd.get_dummies(df['mau'])

False  True 
0         1      0
1         1      0
2         1      0
3         1      0
4         1      0
...     ...    ...
6765      0      1
6766      0      1
6767      0      1
6768      0      1
6769      0      1

[6770 rows x 2 columns]

#### 2) Vamos montar um metadados

1. Crie um dataframe com os nomes de cada variável e o tipo de dados de cada variável.
2. Adicione uma coluna nesse *dataframe* chamada "qtd_categorias" e coloque nela o número de categorias correspondente de cada variável. 
    Dica: 
        1. inicie uma lista vazia
        2. faça um for ao longo dos nomes das variáveis, 
        3. conte o número de categorias dessa variável 
        4. acumule essa informação de 3. na lista que você criou em 1. 
        5. No final, essa lista pode ser atribuída à nossa variável.
3. Crie variáveis dummy para as variáveis necessárias (i.e. aquelas que são qualitativas e não estão armazenadas como {0, 1} ou {True, False}.

* As variáveis 'idade' e 'tempo de emprego' estão muito discrepantes, vamos tratar elas antes continuarmos com as análises


In [218]:
df.head()

index sexo posse_de_veiculo posse_de_imovel  qtd_filhos  \
0      0    M                Y               Y           0   
1      1    F                N               Y           0   
2      3    M                Y               Y           0   
3      4    F                Y               N           0   
4      6    F                N               Y           0   

             tipo_renda                       educacao          estado_civil  \
0               Working  Secondary / secondary special               Married   
1  Commercial associate  Secondary / secondary special  Single / not married   
2               Working               Higher education               Married   
3               Working              Incomplete higher               Married   
4               Working  Secondary / secondary special               Married   

     tipo_residencia      idade  tempo_emprego  possui_celular  \
0  House / apartment  58.832877       3.106849               1   
1  House / apartment  52.356164       8.358904               1   
2  House / apartment  46.224658       2.106849               1   
3  House / apartment  29.230137       3.021918               1   
4  House / apartment  27.482192       4.024658               1   

   possui_fone_comercial  possui_fone  possui_email  qt_pessoas_residencia  \
0                      0            0             0                    2.0   
1                      0            1             1                    1.0   
2                      1            1             1                    2.0   
3                      0            0             0                    2.0   
4                      0            1             0                    2.0   

     mau  
0  False  
1  False  
2  False  
3  False  
4  False

In [219]:
df.nunique()

index                    6770
sexo                        2
posse_de_veiculo            2
posse_de_imovel             2
qtd_filhos                  8
tipo_renda                  5
educacao                    5
estado_civil                5
tipo_residencia             6
idade                    5298
tempo_emprego            3005
possui_celular              1
possui_fone_comercial       2
possui_fone                 2
possui_email                2
qt_pessoas_residencia       9
mau                         2
dtype: int64

* Como visto acima, elas estão fracionadas, vamos primeiramente tratar esse problema

In [220]:
df['idade'] = round(df['idade'], 0)
df['tempo_emprego'] = round(df['tempo_emprego'], 0)

In [221]:
df.nunique()

index                    6770
sexo                        2
posse_de_veiculo            2
posse_de_imovel             2
qtd_filhos                  8
tipo_renda                  5
educacao                    5
estado_civil                5
tipo_residencia             6
idade                      48
tempo_emprego              44
possui_celular              1
possui_fone_comercial       2
possui_fone                 2
possui_email                2
qt_pessoas_residencia       9
mau                         2
dtype: int64

In [222]:
display(df[['idade']].value_counts())
display(df[['tempo_emprego']].value_counts())

idade
41.0     211
40.0     208
37.0     206
39.0     206
34.0     203
42.0     198
35.0     192
30.0     190
44.0     186
29.0     186
28.0     183
32.0     183
33.0     182
38.0     180
52.0     176
45.0     175
54.0     171
47.0     170
43.0     169
51.0     169
31.0     168
55.0     168
36.0     165
48.0     165
60.0     164
59.0     162
46.0     161
49.0     159
50.0     158
27.0     147
56.0     145
57.0     138
53.0     138
58.0     133
61.0     129
62.0     116
63.0     106
64.0     105
26.0      89
65.0      80
25.0      63
66.0      47
67.0      45
24.0      33
68.0      21
23.0      14
22.0       4
69.0       3
dtype: int64

tempo_emprego
-1001.0          1184
 1.0              598
 4.0              513
 3.0              510
 2.0              507
 5.0              462
 6.0              402
 7.0              352
 8.0              310
 9.0              267
 10.0             232
 0.0              178
 11.0             172
 12.0             144
 15.0             126
 13.0             122
 14.0             105
 16.0              67
 17.0              62
 19.0              53
 18.0              51
 20.0              51
 21.0              41
 22.0              37
 23.0              30
 24.0              25
 25.0              23
 26.0              23
 27.0              18
 28.0              18
 31.0              13
 29.0              12
 30.0              12
 33.0              10
 32.0               9
 35.0               7
 34.0               6
 36.0               3
 38.0               3
 39.0               3
 40.0               3
 41.0               3
 37.0               2
 43.0               1
dtype: int64

* A variável 'tempo_emprego' possui um valor de -1001. Vamos cuidar disso

In [223]:
df['tempo_emprego'].value_counts()

-1001.0    1184
 1.0        598
 4.0        513
 3.0        510
 2.0        507
 5.0        462
 6.0        402
 7.0        352
 8.0        310
 9.0        267
 10.0       232
 0.0        178
 11.0       172
 12.0       144
 15.0       126
 13.0       122
 14.0       105
 16.0        67
 17.0        62
 19.0        53
 20.0        51
 18.0        51
 21.0        41
 22.0        37
 23.0        30
 24.0        25
 26.0        23
 25.0        23
 27.0        18
 28.0        18
 31.0        13
 29.0        12
 30.0        12
 33.0        10
 32.0         9
 35.0         7
 34.0         6
 38.0         3
 39.0         3
 40.0         3
 41.0         3
 36.0         3
 37.0         2
 43.0         1
Name: tempo_emprego, dtype: int64

* Substituindo os valores negativos por 0

In [224]:
df['tempo_emprego'].replace(-1001.0, 0, inplace = True)

In [225]:
df['tempo_emprego'].value_counts()

0.0     1362
1.0      598
4.0      513
3.0      510
2.0      507
5.0      462
6.0      402
7.0      352
8.0      310
9.0      267
10.0     232
11.0     172
12.0     144
15.0     126
13.0     122
14.0     105
16.0      67
17.0      62
19.0      53
20.0      51
18.0      51
21.0      41
22.0      37
23.0      30
24.0      25
25.0      23
26.0      23
28.0      18
27.0      18
31.0      13
30.0      12
29.0      12
33.0      10
32.0       9
35.0       7
34.0       6
39.0       3
38.0       3
40.0       3
41.0       3
36.0       3
37.0       2
43.0       1
Name: tempo_emprego, dtype: int64

* Agora seguindo com nosso tratamento dos dados, vamos filtrar os tipos do Data Frame

In [226]:
df_tipo = pd.DataFrame(df.iloc[0:1, 0:].dtypes)

In [227]:
df_tipo.reset_index(inplace = True)

In [228]:
df_tipo.rename(columns = {'index': 'var', 
            0: 'tipo'}, inplace = True)

In [229]:
df_tipo

var     tipo
0                   index    int64
1                    sexo   object
2        posse_de_veiculo   object
3         posse_de_imovel   object
4              qtd_filhos    int64
5              tipo_renda   object
6                educacao   object
7            estado_civil   object
8         tipo_residencia   object
9                   idade  float64
10          tempo_emprego  float64
11         possui_celular    int64
12  possui_fone_comercial    int64
13            possui_fone    int64
14           possui_email    int64
15  qt_pessoas_residencia  float64
16                    mau     bool

* Função para contar o número de categorias das variáveis

In [230]:
def insere_quantidade(valor: pd.DataFrame):
    qtd_categoria = []

    for i in df:
        qtd_categoria.append(df[i].nunique())

    return qtd_categoria

In [231]:
quantidade = insere_quantidade(df)

In [232]:
df_tipo['qtd_categorias'] = quantidade

* Comparativo entre tabela original e tabela com os tipos

In [233]:
display(df.nunique())
display(df_tipo)

index                    6770
sexo                        2
posse_de_veiculo            2
posse_de_imovel             2
qtd_filhos                  8
tipo_renda                  5
educacao                    5
estado_civil                5
tipo_residencia             6
idade                      48
tempo_emprego              43
possui_celular              1
possui_fone_comercial       2
possui_fone                 2
possui_email                2
qt_pessoas_residencia       9
mau                         2
dtype: int64

var     tipo  qtd_categorias
0                   index    int64            6770
1                    sexo   object               2
2        posse_de_veiculo   object               2
3         posse_de_imovel   object               2
4              qtd_filhos    int64               8
5              tipo_renda   object               5
6                educacao   object               5
7            estado_civil   object               5
8         tipo_residencia   object               6
9                   idade  float64              48
10          tempo_emprego  float64              43
11         possui_celular    int64               1
12  possui_fone_comercial    int64               2
13            possui_fone    int64               2
14           possui_email    int64               2
15  qt_pessoas_residencia  float64               9
16                    mau     bool               2

#### 3) Crie variáveis dummy para as variáveis necessárias (i.e. aquelas que são qualitativas e não estão armazenadas como {0, 1} ou {True, False}. Crie um *dataframe* apenas com as variáveis apropriadas para entrada no scikitlearn - elimine as variáveis tipo *str*, mantendo apenas suas versões *dummy*.

In [234]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6770 entries, 0 to 6769
Data columns (total 17 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   index                  6770 non-null   int64  
 1   sexo                   6770 non-null   object 
 2   posse_de_veiculo       6770 non-null   object 
 3   posse_de_imovel        6770 non-null   object 
 4   qtd_filhos             6770 non-null   int64  
 5   tipo_renda             6770 non-null   object 
 6   educacao               6770 non-null   object 
 7   estado_civil           6770 non-null   object 
 8   tipo_residencia        6770 non-null   object 
 9   idade                  6770 non-null   float64
 10  tempo_emprego          6770 non-null   float64
 11  possui_celular         6770 non-null   int64  
 12  possui_fone_comercial  6770 non-null   int64  
 13  possui_fone            6770 non-null   int64  
 14  possui_email           6770 non-null   int64  
 15  qt_p

* Como visto acima, vamos tratar as seguintes variáveis:  
    sexo, posse_de_veiculo, posse_de_imovel, tipo_renda, educacao, estado_civil, tipo_residencia, mau

In [235]:
df[['sexo', 'posse_de_veiculo', 'posse_de_imovel', 'tipo_renda', 'educacao', 'estado_civil', 'tipo_residencia', 'mau']]

sexo posse_de_veiculo posse_de_imovel            tipo_renda  \
0       M                Y               Y               Working   
1       F                N               Y  Commercial associate   
2       M                Y               Y               Working   
3       F                Y               N               Working   
4       F                N               Y               Working   
...   ...              ...             ...                   ...   
6765    F                N               Y               Working   
6766    F                N               Y               Working   
6767    F                N               Y  Commercial associate   
6768    M                Y               Y               Working   
6769    F                N               Y             Pensioner   

                           educacao          estado_civil    tipo_residencia  \
0     Secondary / secondary special               Married  House / apartment   
1     Secondary / secondary special  Single / not married  House / apartment   
2                  Higher education               Married  House / apartment   
3                 Incomplete higher               Married  House / apartment   
4     Secondary / secondary special               Married  House / apartment   
...                             ...                   ...                ...   
6765  Secondary / secondary special  Single / not married  House / apartment   
6766  Secondary / secondary special        Civil marriage  House / apartment   
6767  Secondary / secondary special               Married  House / apartment   
6768  Secondary / secondary special               Married  House / apartment   
6769               Higher education               Married  House / apartment   

        mau  
0     False  
1     False  
2     False  
3     False  
4     False  
...     ...  
6765   True  
6766   True  
6767   True  
6768   True  
6769   True  

[6770 rows x 8 columns]

* sexo

In [236]:
df['sexo'].value_counts()

F    4449
M    2321
Name: sexo, dtype: int64

In [237]:
df['sexo'] = df['sexo'].map({'F': 0, 'M': 1})

In [238]:
df['sexo'].value_counts()

0    4449
1    2321
Name: sexo, dtype: int64

* posse_de_veiculo

In [239]:
df['posse_de_veiculo'].value_counts()

N    4211
Y    2559
Name: posse_de_veiculo, dtype: int64

In [240]:
df['posse_de_veiculo'] = df['posse_de_veiculo'].map({'N': 0, 'Y': 1})

In [241]:
df['posse_de_veiculo'].value_counts()

0    4211
1    2559
Name: posse_de_veiculo, dtype: int64

* posse_de_imovel

In [242]:
df['posse_de_imovel'].value_counts()

Y    4493
N    2277
Name: posse_de_imovel, dtype: int64

In [243]:
df['posse_de_imovel'] = df['posse_de_imovel'].map({'Y': 1, 'N': 0})

In [244]:
df['posse_de_imovel'].value_counts()

1    4493
0    2277
Name: posse_de_imovel, dtype: int64

In [245]:
df['mau'].value_counts()

False    6472
True      298
Name: mau, dtype: int64

In [246]:
df['mau'] = df['mau'].map({False: 0, True: 1})

In [247]:
df['mau'].value_counts()

0    6472
1     298
Name: mau, dtype: int64

* Demais variáveis em dummies

In [248]:
df_encoded = pd.get_dummies(df, columns =  ['tipo_renda', 'educacao', 'estado_civil', 'tipo_residencia'], drop_first = True)

In [249]:
df_encoded.head()

index  sexo  posse_de_veiculo  posse_de_imovel  qtd_filhos  idade  \
0      0     1                 1                1           0   59.0   
1      1     0                 0                1           0   52.0   
2      3     1                 1                1           0   46.0   
3      4     0                 1                0           0   29.0   
4      6     0                 0                1           0   27.0   

   tempo_emprego  possui_celular  possui_fone_comercial  possui_fone  \
0            3.0               1                      0            0   
1            8.0               1                      0            1   
2            2.0               1                      1            1   
3            3.0               1                      0            0   
4            4.0               1                      0            1   

   possui_email  qt_pessoas_residencia  mau  tipo_renda_Pensioner  \
0             0                    2.0    0                     0   
1             1                    1.0    0                     0   
2             1                    2.0    0                     0   
3             0                    2.0    0                     0   
4             0                    2.0    0                     0   

   tipo_renda_State servant  tipo_renda_Student  tipo_renda_Working  \
0                         0                   0                   1   
1                         0                   0                   0   
2                         0                   0                   1   
3                         0                   0                   1   
4                         0                   0                   1   

   educacao_Higher education  educacao_Incomplete higher  \
0                          0                           0   
1                          0                           0   
2                          1                           0   
3                          0                           1   
4                          0                           0   

   educacao_Lower secondary  educacao_Secondary / secondary special  \
0                         0                                       1   
1                         0                                       1   
2                         0                                       0   
3                         0                                       0   
4                         0                                       1   

   estado_civil_Married  estado_civil_Separated  \
0                     1                       0   
1                     0                       0   
2                     1                       0   
3                     1                       0   
4                     1                       0   

   estado_civil_Single / not married  estado_civil_Widow  \
0                                  0                   0   
1                                  1                   0   
2                                  0                   0   
3                                  0                   0   
4                                  0                   0   

   tipo_residencia_House / apartment  tipo_residencia_Municipal apartment  \
0                                  1                                    0   
1                                  1                                    0   
2                                  1                                    0   
3                                  1                                    0   
4                                  1                                    0   

   tipo_residencia_Office apartment  tipo_residencia_Rented apartment  \
0                                 0                                 0   
1                                 0                                 0   
2                                 0                                 0   
3                                 0                                 0   
4          

#### 4) Qual variável é mais poderosa?

Considere as variáveis ```possui_email``` e ```posse_de_veiculo```. Faça uma tabela cruzada entre elas e responda qual delas te parece mais poderosa para prever a probabilidade de ```mau = 1```?

* Como podemos ver, a variável que nos dá uma maior margem de previsão é a 'posse_de_veiculo'

In [250]:
tabela_cruzada = pd.crosstab(df_encoded['possui_email'], df_encoded['posse_de_veiculo'])

In [251]:
tabela_cruzada

posse_de_veiculo     0     1
possui_email                
0                 3878  2327
1                  333   232

#### 5) Salve a base, pois ela será utilizada no final deste módulo.

In [255]:
df_encoded.to_csv(r'demo01_analisada.csv', index = False)